Fix #44 and #45; honor skip_large_bodies while also giving it a default of true

verisk-dhaskin opened the pull request:
This PR fixes skip_large_bodies so that it is honored. Previously it wasn't honored. Because of the way the code was written, if the setting was defined but false, it would be set to true anyway. This new code allows the default to be true but also checks explicitly to see if the setting was defined first, and if it is, sets it according to what was configured.

Typically, in lua, it is idiomatic to write the line

`a = b or some-default`

This works fine if the type of `a` and `b` are *non-boolean* values, since any value that evaluates to `nil` in lua is considered `false` as evaluated in the above expression. This idiom was used here; however, the value is boolean, and so it also evaluates to false if it is defined (non-nil) and is set to false on purpose.

The new code in the pull request explicitly checks for definition of the value separately so that the original meaning of the line is preserved while also honoring any explicitly configured `false` value.

verisk-dhaskin committed:
fix(aws-gw) honor skip_large_bodies for aws-gateway format

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/kong-plugin-aws-lambda?pullRequest=48) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/kong-plugin-aws-lambda?pullRequest=48) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/kong-plugin-aws-lambda?pullRequest=48) it.</sub>

cal5barton commented:
@Tieske any ideas why this change would break the build?

gszr commented:
@verisk-dhaskin Thanks for the PR. This test failure is occurring in `master` as well. We are looking into it!

gszr commented:
@verisk-dhaskin The failure has been fixed; would you please rebase your PR? Thanks!

Tieske commented on kong/plugins/aws-lambda/aws-serializer.lua:
```diff
@@ -54,7 +54,10 @@ return function(ctx, config)
   local body, isBase64Encoded
-  local skip_large_bodies = config and config.skip_large_bodies or true
+  local skip_large_bodies = true
+  if config ~= nil and config.skip_large_bodies ~= nil then
```
```suggestion
  if config and config.skip_large_bodies ~= nil then
```

verisk-dhaskin replied:
I committed your suggestion, but I don't see how it would alter behavior to make the tests fail. I think you meant to take off the `~= nil` against the `config.skip_large_bodies`, not against the `config` part of that clause. `config` is a non-boolean value, so `config ~= nil` is functionally equivalent to `config` in that clause, whereas `config.skip_large_bodies` is a boolean value, so if you had removed the `~= nil` against that, it would have altered the meaning of the line. Is this what you meant to do @Tieske ?

Tieske reviewed (changes requested):
The problem here is not the fix, but the tests not failing. There are tests for this behaviour, but they pass, so first we should fix the tests to properly fail, and then see if this fix then makes them pass again

Tieske committed:
refactor(test) move fixtures to separate file for re-use

verisk-dhaskin commented:
In the above commit, I added a commit that I think will make the tests fail that @Tieske refers to.

verisk-dhaskin commented:
I have made changes that should have failed the tests, but the tests passed. Please review the tests @Tieske , they should have failed.

Tieske committed:
chore(test) add tests for large body setting

This replaces the unit tests with integration tests, and tests both
serialization formats (plain, and aws-gateway compatible)

Tieske commented:
@verisk-dhaskin I took the liberty to update your branch, by adding (and rewriting) tests

I think the reproduction previously failed since the problem only occurs if the plugin is set to be compatible with the AWS-gateway.

Tieske commented:
hmmmm, the log file assertion was only added in 2.x, so I guess we'll have to drop the CI from the 1.3 and 1.5 versions

Tieske commented:
hmmmm, the log file assertion was only added in 2.x, so I guess we'll have to drop the CI from the 1.3 and 1.5 versions

Tieske committed:
chore(CI) drop pre-2.x versions from test matrix

New tests use the logfile assertion not available in pre-2.x
